sqlite3/provider.go

```diff
@@ -0,0 +1,106 @@
+package sqlite3
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/dhermes/golembic"
+)
+
+const (
+	createTableConstraints = `,
+
+  FOREIGN KEY(previous) REFERENCES revision
+  CHECK (
+    (serial_id = 0 AND previous IS NULL) OR
+    (serial_id != 0 AND previous IS NOT NULL)
+  )
+`
+)
+
+// NOTE: Ensure that
+//       * `SQLProvider` satisfies `golembic.EngineProvider`.
+var (
+	_ golembic.EngineProvider = (*SQLProvider)(nil)
+)
+
+// New creates a SQLite-specific database engine provider from some
+// options.
+func New(opts ...Option) (*SQLProvider, error) {
+	cfg := &Config{
+		DataSourceName: DefaultDataSourceName,
+		DriverName:     DefaultDriverName,
+	}
+	for _, opt := range opts {
+		err := opt(cfg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &SQLProvider{Config: cfg}, nil
+}
+
+// SQLProvider is a SQLite-specific database engine provider.
+type SQLProvider struct {
+	Config *Config
+}
+
+// QueryParameter produces the placeholder `?NNN` for a numbered
+// parameter in a SQLite query.
+//
+// See: https://sqlite.org/lang_expr.html#parameters
+func (*SQLProvider) QueryParameter(index int) string {
+	return fmt.Sprintf("?%d", index)
+}
+
+// NewCreateTableParameters produces the SQL expressions used in the
+// the `CREATE TABLE` statement used to create the migrations table.
+func (*SQLProvider) NewCreateTableParameters() golembic.CreateTableParameters {
+	ctp := golembic.NewCreateTableParameters(
+		// H/T: https://stackoverflow.com/a/3693112/1068170
+		golembic.OptCreateTableCreatedAt("INTEGER DEFAULT (CAST((julianday('now') - 2440587.5) * 86400.0 * 1000000 AS INTEGER))"),
+		golembic.OptCreateTableConstraints(createTableConstraints),
+		golembic.OptCreateTableSkip(true),
+	)
+
+	ctp.SerialID += " UNIQUE CHECK (serial_id >= 0)"
+	ctp.Revision += " PRIMARY KEY"
+	ctp.Previous += " UNIQUE CHECK (previous != revision)"
```

Comment:
There is (as of now) nothing that keeps this from drifting when `table.go` changes.